### ISSUE TITLE
Built-in server tools don't trigger postToolUseHook

### ISSUE THREAD

**Issue by andrejmokris** [anthropics/claude-agent-sdk-typescript#101]
### Environment:
  - Claude Agent SDK: 0.1.67
  - Claude Code: 2.0.67 (native install)

 ### Description:
  PostToolUse hook fires correctly for custom MCP tools, but never fires for built-in server tools like WebSearch and WebFetch. Seems to be broken from `0.1.63` release

### Observed behavior:
  1. PreToolUse fires with tool_use_id: "toolu_..."
  2. Tool executes successfully
  3. PostToolUse never fires
  4. PostToolUseFailure never fires
  5. Results appear in user message with tool_use_result field containing different ID prefix (srvtoolu_...)

### Expected behavior:
PostToolUse should fire after WebSearch/WebFetch complete, same as for MCP tools.
Worked correctly on previous versions

**Comment by sleepy-zone** [comment]
and PreToolUseHook ....

**Comment by diagramatics** [comment]
Can confirm rolling back to `0.1.62` fixes this. I looked into `sdk.mjs` and it looks like `handleSingleTurnInputComplete` is the cause. Looks like the `handleSingleTurnInputComplete` logic is too optimistic about `waitForInactivity` when you're just waiting for responses from Claude, even for the first `assistant` message?

`logForDebugging` in `sdk.mjs` immediately returns this before even the first message:
```
2025-12-16T13:13:58.165Z [DEBUG] [Query.handleSingleTurnInputComplete] Has bidirectional needs, waiting for inactivity
2025-12-16T13:13:58.165Z [DEBUG] [Query.waitForInactivity] Waiting for inactivity (timeout: 5000ms)
2025-12-16T13:13:58.165Z [DEBUG] [Query.waitForInactivity] Still active, checking again in 4999ms
2025-12-16T13:14:03.165Z [DEBUG] [Query.waitForInactivity] Still active, checking again in 570ms
2025-12-16T13:14:03.737Z [DEBUG] [Query.waitForInactivity] Inactivity timeout reached (5002ms elapsed). Closing stdin. If your tools or hooks need more time, set CLAUDE_CODE_STREAM_CLOSE_TIMEOUT to a higher value (current: 5000ms).
2025-12-16T13:14:03.738Z [DEBUG] [Query.handleSingleTurnInputComplete] Calling transport.endInput()
```

And this causes the `cli.js` to throw a callback hook error saying the stream is closed.

```
2025-12-16T13:08:02.654Z Error in hook callback hook_2: Error: Stream closed
    at fPA.sendRequest (file:///.../node_modules/@anthropic-ai/claude-agent-sdk/cli.js:4770:117)
    at Object.callback (file:///.../node_modules/@anthropic-ai/claude-agent-sdk/cli.js:4770:1212)
    at RD5 (file:///.../node_modules/@anthropic-ai/claude-agent-sdk/cli.js:4607:1198)
    at file:///.../node_modules/@anthropic-ai/claude-agent-sdk/cli.js:4600:1127
    at AsyncGenerator.next (<anonymous>)
    at B (file:///.../node_modules/@anthropic-ai/claude-agent-sdk/cli.js:1854:2437)
    at QIA (file:///.../node_modules/@anthropic-ai/claude-agent-sdk/cli.js:1854:2590)
    at QIA.next (<anonymous>)
    at Ra (file:///.../node_modules/@anthropic-ai/claude-agent-sdk/cli.js:4600:4169)
    at Ra.next (<anonymous>)
```
This `handleSingleTurnInputComplete` seems to be replacing this particular logic in `readMessages` that is responsible to check for the `result` message and ends it there.

```js
if (message.type === "result") {
  if (this.firstResultReceivedResolve) {
    this.firstResultReceivedResolve();
  }
  if (this.isSingleUserTurn) {
    this.transport.endInput();
  }
}
```

And also this logic to race for the first message received, or to end the stream after a timeout.

```
if ((this.sdkMcpTransports.size > 0 || hasHooks) && this.firstResultReceivedPromise) {
  logForDebugging(`[Query.streamInput] Entering Promise.race to wait for result`);
  let timeoutId;
  await Promise.race([
    this.firstResultReceivedPromise.then(() => {
      logForDebugging(`[Query.streamInput] Received first result, closing input stream`);
      if (timeoutId) {
        clearTimeout(timeoutId);
      }
    }),
    new Promise((resolve) => {
      timeoutId = setTimeout(() => {
        logForDebugging(`[Query.streamInput] Timed out waiting for first result, closing input stream`);
        resolve();
      }, this.streamCloseTimeout);
    })
  ]);
  if (timeoutId) {
    clearTimeout(timeoutId);
  }
}
```

These two exist in `0.1.62` and replaced by `handleSingleTurnInputComplete` on `0.1.63`.

Interestingly, `0.1.64` [claims to have fixed this](https://github.com/anthropics/claude-agent-sdk-typescript/blob/main/CHANGELOG.md#0164) but I can't find the release published on `npm` anymore, and `0.1.65` doesn't seem to have the fix?

**Comment by Avadis7860** [comment]
We're observing the same broken `PostToolUse` hook behavior described in this issue on SDK `@anthropic-ai/claude-agent-sdk@0.2.139`, but the scope appears **extended to custom MCP HTTP tools** (not just built-in hosted-Anthropic ones).

**Reproduction reliability**: consistently reproduced across multiple smoke runs over two debug cycles (5+ runs total), with identical behavior under three different hook matcher configurations (catchall / specific exact-match / regex) and two `settingSources` configurations (`["project"]` and exhaustive `["user","project","local"]`).

**Setup** (TypeScript + Node 20):
- Hook: `PostToolUse` with specific matcher `"mcp__vault__vault_search"` (underscore form)
- Custom MCP server transport: `type: "http"`, remote HTTPS endpoint
- Built-in tools (`Glob`, `Read`, `Bash`) DO fire `PostToolUse` correctly — only the MCP HTTP tool is silently skipped
- The tool IS being called by Claude (confirmed via in-loop `tool_use` content block inspection), but `PostToolUse` never fires for it

**Workaround we deployed**: inline `INSERT` discipline in the for-await SDK message loop (branches `tool_use` assistant msg + `tool_result` user msg), with a dedup `Set` keyed by `tool_use_id` to prevent double-INSERT if the hook ever starts firing upstream — keeps hook registered for future-proofing zero-cost.

Suggesting the issue scope be extended to **custom MCP HTTP tools**, not just hosted Anthropic built-ins (`srvtoolu_*` WebSearch/WebFetch). Happy to provide deeper repro details (container logs, commit refs) if useful for maintainer triage.

**Comment by Avadis7860** [comment]
Confirming this issue bidirectionally — PostToolUse hooks do not fire on SDK 0.2.139 for BOTH:

1. **Built-in tools** (`srvtoolu_*`)
2. **Custom MCP servers over HTTP Streamable transport**

Reproduced empirically in our agent-runner project after 3 diagnostic cycles (G11 → F6 → G12). Tried:

- Matcher with proper underscore form (`mcp__<server>__<tool_underscore>`) — Anthropic API tool name regex `^[a-zA-Z0-9_-]{1,128}$` forbids `.`, SDK rewrites `vault.search` → `vault_search` before sending.
- Matcher-less catch-all logger — 0 fire.
- `settingSources: ["user", "project", "local"]` exhaustive — 0 fire.
- `PreToolUse` hook on the same matcher — 0 fire (suggests upstream of fire-policy, not matcher logic).

Workaround: in-process tracking from the `query()` for-await loop, joining `assistant.tool_use` blocks (by `tool_use_id` / `toolu_*`) with `user.tool_result` blocks, then persisting / instrumenting from the application layer. Future-proofed with a shared dedup `Set<string>` keyed by `tool_use_id` in case the SDK eventually fixes hook fire-policy.

Would be great to get clarity on whether PostToolUse is intended to fire for HTTP MCP custom tools and built-ins, or if this is a documented gap. The README phrasing implies general coverage but empirical behavior says otherwise.
